### DISCUSSION TITLE
How do I use DNS Service Discovery inside quarkus?

### DISCUSSION THREAD

**Discussion by rjsoph** [smallrye/smallrye-stork#386] [Q&A]
I am trying to get DNS Service Discovery working but i get this error...

`Failed to start application (with profile prod): java.lang.AbstractMethodError: Receiver class io.smallrye.stork.servicediscovery.dns.DnsServiceDiscoveryProviderLoader does not define or inherit an implementation of the resolved method 'abstract io.smallrye.stork.api.ServiceDiscovery createServiceDiscovery`

I have included

```
<dependency>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-smallrye-stork</artifactId>
</dependency>
<dependency>
  <groupId>io.smallrye.stork</groupId>
  <artifactId>stork-service-discovery-dns</artifactId>
  <version>1.3.0</version>
</dependency>
```

and have my properties setup

```
quarkus.stork.stripe-service.service-discovery.type=dns
quarkus.rest-client."com.xxxx.clients.stripe.StripeService".url=stork://stripe-service/micro
```

feels like i am missing a dependency.....  or do I need to write a custom provider?

**Comment by aureamunoz** [comment]
Hi,

I don't see any particular issue, your config seems ok.

Let me try to reproduce the error and see how it goes

**Reply by cescoffier** [reply]
Which version of Quarkus are you using? 
In general, do not set the version for stork dependencies as they are aligned by quarkus.

**Reply by rjsoph** [reply]
I am or Quarkus is?  I am using Quarkus 2.13.0.Final and including quarkus-smallrye-stork dependency.  Is there a version of the Quarkus distribution I can use or do I need to try to implement this outside Quarkus?

**Reply by rjsoph** [reply]
I am not specifying the version my dependencies are as shown above in original post.

**Comment by aureamunoz** [comment]
Yes, it seems like you are using an old stork version and it is not generating the `DnsServiceDiscoveryProviderLoader.java` class

**Reply by rjsoph** [reply]
I am using the one in the Quarkus 2.13.0.Final.... I see they are up to 2.13.2.... I will try to upgrade.

**Reply by rjsoph** [reply]
The upgrade to 2.13.2 did not fix the problem.

**Reply by cescoffier** [reply]
Can you share a reproducer? I've worked with @aureamunoz on a stork dns example this afternoon and it was working as expected.

**Reply by rjsoph** [reply]
For me just including the quarkus.stork....service-discover.type=dns prevents the application from starting up.   If I use anything else or comment out service discovery altogether is starts up fine.  Was your working example on Quarkus?  I feel like its an issue with the Quarkus stork distribution.

**Reply by aureamunoz** [reply]
Quarkus 2.13.x uses Stork 1.1.2 that doesn't provide the Dns Service Discovery that's the problem.

In the original post there is 1.3.0 for stork dns dependency but it has not effect because is Quarkus bom who fixes this version

**Reply by aureamunoz** [reply]
Yes, we work in a dns example with Quarkus where we override the Stork version to use latest 1.3.0 and that works

**Reply by rjsoph** [reply]
Can you help me with how to do that?  When I add a version to the quarkus stork dependency it complains about not being able to find the jar.

**Reply by rjsoph** [reply]
I think I got it working by avoiding the quarkus-stork and using smallrye.... then changing the property accordingly.

Thank you both!

**Reply by aureamunoz** [reply]
Next week I will push the example in the Quarkus Quickstart repo but it will use the Quarkus main branch, not a final release. In the meantime can you try to use the Stork as standalone instead of using the Quarkus extension? This way you can use the latest version...

**Reply by aureamunoz** [reply]
ah, just read your last comment, I hadn't seen it when writing my last. So you used it as standalone and it works, right? 
I'll let you know when push the dns example using Quarkus

**Reply by rjsoph** [reply]
I spoke too soon... i seem to need the quarkus-stork libraries to Register my clients with "stork" protocol.... and including stork-core and stork-dns from smallrye does not seem to be enough.... I through in stork-api but that seems to be leading me down a rabbit hole with no end in sight...   can you give me a dependency tree for a working quarkus example?

**Reply by rjsoph** [reply]
The stork documentation claims DNS support in 1.1.2....  [https://smallrye.io/smallrye-stork/1.1.2/service-discovery/dns/](https://smallrye.io/smallrye-stork/1.1.2/service-discovery/dns/ ).

I tried to switch to "static" as a work around and it seems to be plagued by the same issue....

**Reply by cescoffier** [reply]
Can you provide a reproducer? 

There is also a difference between the upstream version and the version supported in Quarkus.

**Reply by cescoffier** [reply]
Btw, stork does not handle the registration of your service into your DNS server. You need to do that yourself (depending on your platform, in could be done automatically (Kubernetes for example)

**Reply by rjsoph** [reply]
I am using AWS cloud map... it takes care of registration with DNS.  
[example.zip](https://github.com/smallrye/smallrye-stork/files/9789521/example.zip)

**Reply by rjsoph** [reply]
I included the wrong properties file the class name for my Rest Client is wrong..... should be com.example.StripeService.  It does not matter, the error happens upon startup never gets far enough to connect....

**Reply by rjsoph** [reply]
I am beginning to suspect it may be because my application is not reactive.... is that a requirement for stork?  Seems like an odd way to enforce that.....

**Reply by aureamunoz** [reply]
Hi @rjsoph !

I've been able to reproduce your error by using Quarkus **2.13.2.Final** also in our example. The is that Stork 1.1.2 doesn't support DNS service discovery. I've just opened [an issue](https://github.com/smallrye/smallrye-stork/issues/391) to fix the documentation part.

So as Quarkus 2.13.2.Final is using an old Stork version, it won't work. For having DNS service discovery, use Stork 1.3.0. This is possible by using Quarkus 999-SNAPSHOT which is built nightly. For that, add the following in your pom:

```
<repositories>
        <repository>
            <id>ossrh</id>
            <url>https://s01.oss.sonatype.org/content/repositories/snapshots/</url>
            <snapshots>
                <enabled>true</enabled>
                <updatePolicy>always</updatePolicy>
            </snapshots>
        </repository>
    </repositories>
    <pluginRepositories>
        <pluginRepository>
            <id>ossrh</id>
            <url>https://s01.oss.sonatype.org/content/repositories/snapshots/</url>
            <snapshots>
                <enabled>true</enabled>
                <updatePolicy>always</updatePolicy>
            </snapshots>
        </pluginRepository>
    </pluginRepositories>
```

and 

```
<quarkus.platform.version>999-SNAPSHOT</quarkus.platform.version>
```
Also, as already mentioned, don't set set the version for stork dependencies as they are aligned by quarkus, just the `groupId` and `artifactId`:

```
<dependency>
      <groupId>io.smallrye.stork</groupId>
      <artifactId>stork-service-discovery-dns</artifactId>
</dependency>
```
And that should work.

You can check our example [here](https://github.com/aureamunoz/quarkus-quickstarts/tree/stork-dns/stork-dns-quickstart).

Let me know if it works.

**Reply by rjsoph** [reply]
I am having other issues with linking my packages using the 999-SNAPSHOT that I would rather not chase down.  E.g. Hibernate ConfigurationService seems to be missing.  

How long before this makes its way into an official release?  Trying to weigh my deployment options...

**Reply by aureamunoz** [reply]
Try using 
```
<quarkus.platform.group-id>io.quarkus</quarkus.platform.group-id>
```
